### PR TITLE
Add `@model_validator`

### PR DESCRIPTION
**Pull request by adriangb**
Selected Reviewer: @hramezani

### CODE REVIEW

File: pydantic/_internal/_decorators.py [hunk 1]
```diff
@@ -786,3 +803,19 @@ def wrap_model_serializer_in_wrap_mode(
             raise TypeError(f'Unrecognized serializer signature for {serializer} with `mode={mode}`:{sig}')
         func = cast(AnyCoreSerializer, serializer)
         return func
+
+
+ModelType = TypeVar('ModelType', bound='BaseModel')
```
Comment (dmontagu) [inline pydantic/_internal/_decorators.py:808]: Should we loosen the bound here? E.g., I'm thinking this might apply to a dataclass or whatever.

**Review by dmontagu — COMMENTED** [review]
Generally looking good to me

File: pydantic/_internal/_decorators.py (hunk 1 above)
Comment (adriangb) [reply pydantic/_internal/_decorators.py:808]: Maybe, but then we end accepting `Callable[[Any, Any], Any]` in some situations lol. It's kind of a struggle to type these decorators, especially given how we let you include `info` or not and how we accept `classmethod[Any]` and such. I almost want to sit down with some of the typing folks at PyCon and see if they have any ideas. I do plan to do a bit more typing work on this PR, so I can play around with loosening the bound.

File: pydantic/_internal/_decorators.py [hunk 2]
```diff
@@ -191,12 +193,32 @@ def __init__(
...
+class ModelValidatorDecoratorInfo(Representation):
+    """
+    A container for data from `@model_validator` so that we can access it
+    while building the pydantic-core schema.
+    """
+
+    __slots__ = ('mode',)
+
+    mode: Literal['wrap', 'before', 'after']
+    json_return_type: JsonReturnTypes | None
```
Comment (samuelcolvin) [inline pydantic/_internal/_decorators.py:205]: can be removed.

As mentioned on #5379, I think we should convert these to dataclasses.

File: pydantic/_internal/_decorators.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:808]: I think we need to relax it to at least accept dataclass.

On this, I'd rather start loose and move to stricter types.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM. please update.

File: pydantic/_internal/_decorators.py (hunk 1 above)
Comment (adriangb) [reply pydantic/_internal/_decorators.py:805]: relaxed

File: pydantic/_internal/_decorators.py (hunk 2 above)
Comment (adriangb) [reply pydantic/_internal/_decorators.py:210]: I'm confused: you linked to this pull request

Comment (adriangb) [reply pydantic/_internal/_decorators.py:210]: Nvm, makes sense now. See https://github.com/pydantic/pydantic/pull/5416

**Comment by adriangb** [comment]
please review

**Comment by dmontagu** [comment]
LGTM

**Pull request merged**